lib/src/header_parser/sub_parsers/macro_parser.dart

```diff
@@ -131,17 +132,41 @@ int _macroVariablevisitor(Pointer<clang_types.CXCursor> cursor,
           );
           break;
         case clang_types.CXEvalResultKind.CXEval_StrLiteral:
-          var value = Utf8.fromUtf8(clang.clang_EvalResult_getAsStr(e).cast());
-          // Escape $ character.
-          value = value.replaceAll(r'$', r'\$');
-          // Escape ' character, because our strings are enclosed with '.
-          value = value.replaceAll("'", r"\'");
+          final strPtr = clang.clang_EvalResult_getAsStr(e); // Do not free.
+          final length = Utf8.strlen(strPtr.cast());
+          final charList = Uint8List.view(
+              strPtr.cast<Uint8>().asTypedList(length).buffer, 0, length);
+
+          final sb = StringBuffer();
+          for (final char in charList) {
+            if (char >= 32 && char <= 126) {
+              // Handle the printable range of ASCII.
+              final printableChar = String.fromCharCode(char);
+              // Escape $ and ' character.
+              if (printableChar == '\$' ||
+                  printableChar == "'" ||
+                  printableChar == '\\') {
+                sb.write('\\$printableChar');
+              } else {
+                sb.write(printableChar);
+              }
+            } else {
+              /// Range [127..255] are extended ASCII and [0..31, 127] are
+              /// Control characters. Extended ASCII set cannot be decoded by
+              /// dart:convert and the Control characters cannot be printed
```

Comment:
`dart:convert` has https://api.dart.dev/stable/2.9.3/dart-convert/AsciiDecoder-class.html a. Does the AsciiDecoder not work for extended ascii?

Reply:
It does not, unless i do `AsciiDecoder(allowInvalid: true);`.
Even so, because we need to write the characters to a file to be used, we cannot simply print it.
E.g `\x80` is printed as  (it seems to be showing a question mark in GitHub preview) which I suppose isn't something we write as it is in out generated file

Reply:
```
FormatException: Invalid value in input: 128
[SEVERE] : #0      _UnicodeSubsetDecoder.convert (dart:convert/ascii.dart:178:11)
```

Reply:
> do we know which for of extended ascii clang is giving us through the API

I am not sure what you mean by this, Clang is simply returning a char* which we need to work with now. `Utf8.toUtf8` from `package:ffi` is also crashing, I suppose it should be handled as well.

Reply:
`char*` in C could be any encoding, it could be ascii, extended ascii, utf8. Once you try to decode with the wrong expected encoded you get failures as the ones you mentioned. See https://en.wikipedia.org/wiki/Mojibake for the source of replacement characters.

The way to prevent decoding errors is to know what encoding was used to create the `char*` (the list of bytes).

Reply:
Oh, I get it.

So for this file (1)
```C
#define TEST "this is a string"
```
The file we generate for parsing this is (2)
```cpp
auto _4_TEST_generated_macro_variable = TEST;
```
and clang returns the value of this variable to us.

Now the question is, does the encoding depend on the file encoding where `TEST` is declared (1)? or does it depend on the encoding used for the file we generated (2)?